File: algorithms/Valid Perfect Square/Valid Perfect Square.cpp
class Solution {
public:
	bool isPerfectSquare(int num) {

		if (num == 1)return true;
		long long l = 1, h = num / 2;

		while (l <= h) {
			long long mid = l + (h - l) / 2;
			long long midSqr = mid * mid;

			if (midSqr == num) return true;

			if (num < midSqr) {
				h = mid - 1;
			} else {
				l = mid + 1;
			}
		}
		return false;
	}
};

File: algorithms/V/Valid Perfect Square/Valid Perfect Square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        beg, end = 0, 65535
        
        while beg <= end:
            mid = (beg+end)//2
            if mid*mid == num:
                return True
            elif mid*mid < num:
                beg =  mid+1
            else:
                end = mid-1
        return False
